### DISCUSSION TITLE
Baggage add/update in Span Processor is not working

### DISCUSSION THREAD

**Discussion by srinusanduri** [open-telemetry/opentelemetry-java-instrumentation#6942] [Q&A]
Hello,

We have this requirement to add or update a baggage value at the start of span. We want to add baggage if it's absent, and if the baggage is present, update it if required on each span start. Each span uses the baggage added/updated by previous span and then updates the baggage if required with a new value. We tried using SpanProcessor to achieve this but this is not working as expected. The baggages added or updated by span processor are not accessible to following spans in span processor within the same service. We have tried adding baggages by couple ways as listed below.

```
         final Context context = Context.current();
         context.with(Baggage.current().toBuilder().put(key, value).build()).makeCurrent();
```

```
        BaggageBuilder builder = Baggage.current().toBuilder();
        for (Map.Entry<String, BaggageEntry> entry : Baggage.fromContext(Context.current()).asMap().entrySet()) {
            builder = builder.put(entry.getKey(), entry.getValue().getValue());
        }
        Context.current().with(builder.put(key, value).build()).makeCurrent();
```

Baggage add/update is working fine if we add above code in the service (manual instrumentation using sdk). But it's not working from span processor.

Is it possible to achieve baggage manipulation from Span Processor? If not, are there any alternative ways of achieving this use case? Please provide your suggestions. Thank you.

**Comment by trask** [comment]
hi @srinusanduri! ya, [SpanProcessor](https://javadoc.io/doc/io.opentelemetry/opentelemetry-sdk-trace/latest/io/opentelemetry/sdk/trace/SpanProcessor.html) doesn't provide a way to return an updated Context.

opentelemetry-instrumentation-api provides a [ContextCustomizer](https://javadoc.io/doc/io.opentelemetry.instrumentation/opentelemetry-instrumentation-api/latest/io/opentelemetry/instrumentation/api/instrumenter/ContextCustomizer.html), but there's not a way to hook up a global ContextCustomizer (at least not yet), plus it only works for instrumentation which uses the Instrumenter API, and won't work for instrumentation which uses base opentelemetry-api

**Reply by srinusanduri** [reply]
Thank you for the response @trask. Looks like there is no plugin to update the context as of today. We are migrating from OpenTracing to OpenTelemetry. [OpenTracing Span API](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/Span.java#L137) has the interface to set baggage after the span is started and expected this feature to be already available in OpenTelemetry as well. Ability to add/update baggage to context on span start by auto-instrumenter is a basic common feature. 

I see that there is an [open issue](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/6651) you have created to achieve this functionality using global ContextCustomizer. Hopefully this issue gets resolved soon.

Isn't it possible to add span context update functionality in SpanProcessor itself? One would want to compute a baggage value in span processor and then add/update the computed value in span processor itself.

**Reply by trask** [reply]
@carlosalberto have you run across this migration issue before?
